### ISSUE TITLE
Firebase Database on orderByChild / orderByKey returns in random order

### ISSUE THREAD

**Issue by kroikie** [FirebaseExtended/flutterfire#753]
`firebase_database: 1.0.2`

## Steps to Reproduce

Run a simple Flutter app:
https://gist.github.com/dotdoom/5180eb27084afb3b71db44f2201c6877

Most important part is (for history purposes):

```dart
(await FirebaseDatabase.instance
            .reference()
            .child(subpath)
            .orderByChild('year')
            .limitToFirst(3)
            .onValue
            .first)
        .snapshot
        .value
        .values
        .forEach((v) => debugPrint(v['year'].toString()));
```

The output is indeed the first 3 entries of a sorted list, but the entries themselves are not sorted:

```
I/flutter ( 6096): Sorted years:
I/flutter ( 6096): 1856
I/flutter ( 6096): 1867
I/flutter ( 6096): 1564
```

`limitToFirst` does not affect the order and is only to demonstrate that Firebase server gets the correct request.

I suspect there might be some issue when [transferring data between Java and Dart](https://github.com/flutter/plugins/blob/cd4829a6cf3afcf687fd064f18a028b519644d66/packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java#L185) where the map ordering is lost. Unfortunately it's buried too deep for me to find quickly, so I hope someone with more experience in the MethodChannel mechanism can help me out. The type of the map in Flutter app is `_InternalLinkedHashMap` which IIUC is supposed to be ordered.

## Logs

```
[✓] Flutter (Channel beta, v0.5.1, on Linux, locale en_US.UTF-8)
    • Flutter version 0.5.1 at /home/artem/pkg/flutter
    • Framework revision c7ea3ca377 (7 weeks ago), 2018-05-29 21:07:33 +0200
    • Engine revision 1ed25ca7b7
    • Dart version 2.0.0-dev.58.0.flutter-f981f09760

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.0)
    • Android SDK at /home/artem/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.0
    • Java binary at: /home/artem/pkg/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 3.1)
    • Android Studio at /home/artem/pkg/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[!] VS Code (version 1.25.1)
    • VS Code at /usr/share/code
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected devices (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.1.0 (API 27) (emulator)

! Doctor found issues in 1 category.
```

**Comment by BondarenkoStas** [comment]
I think it will be ok to re-open the issue basing on https://github.com/flutter/flutter/issues/19389#issuecomment-506513807.

**Comment by jorgevvr** [comment]
I have this problem still. Trying order and not working.

**Comment by dotdoom** [comment]
tl;dr: this is WAI, see #711.

The **fact** is that Flutter's [`DataSnapshot.value`](https://pub.dev/documentation/firebase_database/latest/firebase_database/DataSnapshot/value.html) is supposed to be an API equivalent of Android [`DataSnapshot.getValue`](https://firebase.google.com/docs/reference/android/com/google/firebase/database/DataSnapshot#public-object-getvalue-boolean-useexportformat), which "returns the data contained in this snapshot as *native types*" (emphasis mine).

Now the **problem** is that "native type" for Map in Java is HashMap, which is unordered. So this is WAI because Flutter team is making efforts to mirror native API as closely as possible.

To **solve** the problem of ordering, Flutter plugin has to provide an API to iterate over children in order (much like in Java one would do with [`DataSnapshot.getChildren`](https://firebase.google.com/docs/reference/android/com/google/firebase/database/DataSnapshot#public-iterabledatasnapshot-getchildren-)). This is being requested in #711.

**Comment by GursheeshSingh** [comment]
Having the same problem, is there any way to fix this?

**Comment by alona1227** [comment]
not yet

**Comment by hakondh** [comment]
I am experiencing this problem as well. Still no fix?

**Comment by ZiadMontaser** [comment]
Still experiencing the same issue

**Comment by moksh-mahajan** [comment]
Hit by this issue today :(

**Comment by diegogarciar** [comment]
For those having the problem, the fastest workaround for now is to get the keys of the queried documents, call sort() on the keys and and do your loop with those sorted keys. This is possible because database keys are generated time wise, so sorting them actually sorts them from oldest to newest.

**Comment by puf** [comment]
It sounds like this issue should be reopened in that case: https://github.com/FirebaseExtended/flutterfire/issues/711

**Comment by Mayb3Nots** [comment]
having this issue too

**Comment by puf** [comment]
Until this issue is fixed, you can get the keys in correct order by listening to the `onChild...` streams.

So:
```
(await FirebaseDatabase.instance
            .reference()
            .child(subpath)
            .orderByChild('year')
            .limitToFirst(3)
            .onValue
            .first)
        .onChildAdded
        .listen((Event event) {
          print(event.snapshot.value['year']);
        })
```

There is a great utility class called `FirebaseList` included in the package that makes this much easier:
```
list = FirebaseList(query: query, 
  onChildAdded: (pos, snapshot) {},
  onChildRemoved: (pos, snapshot) {},
  onChildChanged: (pos, snapshot) {},
  onChildMoved: (oldpos, newpos, snapshot) {},
  onValue: (snapshot) {
    for (var i=0; i < this.list.length; i++) {
      print('$i: ${list[i].value['year']}');
    }
  }
);
```

In the `onValue` here you can access `this.list` as an array and the child nodes from `query` are present in the correct order in.

Also see my answers to questions about this topic on Stack Overflow:

* https://stackoverflow.com/questions/51808719/flutter-firebase-database-wrong-timestamp-order/51811808#51811808
* https://stackoverflow.com/questions/61333194/flutter-firebase-real-time-database-orderbychild-has-no-impact-on-query-result/61333575#61333575

**Comment by Salakar** [comment]
@puf thanks for posting a workaround.

cc @russellwheatley @rrousselGit I know Database isn't one of the plugins in our purview at the moment but supporting `forEach` looks like something that could have a lot of positive impact here; so we could try look at adding it in ourselves and model it after our implementation in React Native Firebase. Let's discuss feasibility internally.
